examples/text-input.c

```diff
@@ -0,0 +1,393 @@
+\n\
+The \"width\" and \"height\" arguments define the window shape.\n\
+\n\
+The console will print the internal state of the text field:\n\
+- the text in the 1st line\n\
+- \".\" under each preedit character\n\
+- \"_\" under each selected preedit character\n\
+- \"|\" at the cursor position if there are no selected characters in the\n\
+preedit.\n\
+\n\
+The cursor positions may be inaccurate, especially in presence of zero-width\n\
+characters or non-monospaced fonts.\n";
+
+struct text_input_state {
+	char *commit;
+	struct {
+		char *text;
+		int32_t cursor_begin;
+		int32_t cursor_end;
+	} preedit;
+	struct {
+		uint32_t after_length;
+		uint32_t before_length;
+	} delete_surrounding;
+};
+
+static struct text_input_state pending = {0};
+static struct text_input_state current = {0};
+static bool entered = false;
+static uint32_t serial;
+static char *buffer; // text buffer
+// cursor is not present, there's no way to move it outside of preedit
+
+static int sleeptime = 0;
+static int width = 100, height = 200;
+static int enabled = 0;
+
+static struct wl_display *display = NULL;
+static struct wl_compositor *compositor = NULL;
+static struct wl_seat *seat = NULL;
+static struct xdg_wm_base *wm_base = NULL;
+static struct zwp_text_input_manager_v3 *text_input_manager = NULL;
+static struct zwp_text_input_v3 *text_input	= NULL;
+
+struct wlr_egl egl;
+struct wl_egl_window *egl_window;
+struct wlr_egl_surface *egl_surface;
+
+static void draw(void) {
+	eglMakeCurrent(egl.display, egl_surface, egl_surface, egl.context);
+
+	float color[] = {1.0, 1.0, 0.0, 1.0};
+	color[0] = enabled * 1.0;
+	color[1] = entered * 1.0;
+
+	glViewport(0, 0, width, height);
+	glClearColor(color[0], color[1], color[2], 1.0);
+	glClear(GL_COLOR_BUFFER_BIT);
+
+	eglSwapBuffers(egl.display, egl_surface);
+}
+
+static size_t utf8_strlen(char *str) {
+	size_t cp_count = 0;
+	for (; *str != '\0'; str++) {
+		if ((*str & 0xc0) != 0x80) {
+			cp_count++;
+		}
+	}
+	return cp_count;
+}
+
+static size_t utf8_offset(char *utf8_str, size_t byte_offset) {
+	size_t cp_count = 0;
+	for (char *c = utf8_str; c < utf8_str + byte_offset; c++) {
+		if ((*c & 0xc0) != 0x80) {
+			cp_count++;
+		}
+	}
+	return cp_count;
+}
+
+// TODO: would be nicer to have this text display inside the window
+static void show_status() {
+	printf("State %d:", serial);
+	if (!enabled) {
+		printf(" disabled");
+	}
+
+	char *preedit_text = current.preedit.text;
+	if (!preedit_text) {
+		preedit_text = "";
+	}
+
+	printf("\n");
+	printf("%s", buffer);
+	printf("%s\n", preedit_text);
+
+	// Positioning of the cursor requires UTF8 offsets to match monospaced
+	// glyphs
+	for (unsigned i = 0; i < utf8_strlen(buffer); i++) {
+		printf(" ");
+	}
+	char *cursor_mark = calloc(utf8_strlen(preedit_text) + 2, sizeof(char));
+	for (unsigned i = 0; i < utf8_strlen(preedit_text); i++) {
+		cursor_mark[i] = '.';
+	}
+	if (current.preedit.cursor_begin == -1
+			&& current.preedit.cursor_end == -1) {
```

Comment:
cursor_mark leaks here and at the other returns (where we don't goto end). Found by clang https://builds.sr.ht/~emersion/job/8183#task-clang-334 .

Reply:
Fixed